mmwsts.c

```diff
@@ -0,0 +1,1069 @@
+  int ret = strcmp(key, g_MathToken[ *((long *)data) ].tokenName);
+  if(ret != 0) return ret;
+
+  /* We have found two identical entries.
+   * However we shall return 0 (match) only if the token is global. */
+  if(g_MathToken[ *((long *)data) ].endStatement == g_statements) return 0;
+
+  /* Find the direction in which the target token is */
+  for(long *ptr = (long*)data; !strcmp(key, g_MathToken[ *ptr ].tokenName); ptr++)
+    if(g_MathToken[ *ptr ].endStatement == g_statements) return 1;
+  return -1;
+}
+
+/* Similar to mmpars's parseMathTokens, this is a general-purpose
+ * function to parse a math token string. However this version does
+ * not allow to skip spaces. */
+nmbrString *parseMathStrict(vstring text)
+{
+  /* The result */
+  nmbrString *mathString = NULL_NMBRSTRING;
+  long mathStringLen = 0;
+
+  /* Temporary working space */
+  long wrkLen;
+  nmbrString *wrkNmbrPtr;
+
+  long i;
+  char *fbPtr;
+  long textLen, tokenLen_;
+  long *g_mathKeyPtr; /* bsearch returned value */
+
+  /* Make sure that g_mathTokens has been initialized */
+  if (!g_mathTokens) bug(1717);
+
+  textLen = (long)strlen(text);
+
+  /* Initialize temporary working space for parsing tokens */
+  /* Assume the worst case of one token per 2 text characters */
+  wrkLen = textLen/2+1;
+  wrkNmbrPtr = malloc((size_t)wrkLen * sizeof(nmbrString));
+  if (!wrkNmbrPtr) outOfMemory("#502 (wrkNmbrPtr)");
+
+  /* Scan the math section for tokens */
+  fbPtr = text;
+  while (1) {
+    fbPtr = fbPtr + whiteSpaceLen(fbPtr);
+    tokenLen_ = tokenLen(fbPtr);
+    if (!tokenLen_) break; /* Done scanning source line */
+
+    fbPtr[tokenLen_] = 0; /* End token - There's no come back */
+    //printf("Searching for %s\n", fbPtr);
+    g_mathKeyPtr = (long *)bsearch(fbPtr, g_mathKey,
+	(size_t)g_mathTokens, sizeof(long), mathSrchGlbCmp);
+    if (!g_mathKeyPtr) {
+      /* Unknown token, escape. */
+      print2("?Unknown token %s\n", fbPtr);
+      free(wrkNmbrPtr);
+      return NULL_NMBRSTRING;
+    }
+    wrkNmbrPtr[mathStringLen] = *g_mathKeyPtr;
+    mathStringLen++;
+    fbPtr = fbPtr + tokenLen_ + 1; /* Move on to next token */
+    if(fbPtr >= text + textLen) break;
+  }
+
+  /* Assign mathString */
+  nmbrLet(&mathString, nmbrSpace(mathStringLen));
+  for (i = 0; i < mathStringLen; i++)
+    mathString[i] = wrkNmbrPtr[i];
+  if (mathStringLen) nmbrMakeTempAlloc(mathString); /* Flag for dealloc*/
+  free(wrkNmbrPtr);
+  return mathString;
+}
+
+/* Store a couple key/object into the cache */
+void stsStoreCache(nmbrString **tKey, vstring *tObject, nmbrString *sKey, vstring sObject) {
+  *tKey = NULL_NMBRSTRING;
+  nmbrLet(tKey, sKey);
+
+  *tObject = "";
+  let(tObject, sObject);
+}
+
+/* Free a couple key/object from the cache */
+void stsFreeCache(nmbrString **key, vstring *object) {
+  nmbrLet(key, NULL_NMBRSTRING);
+  let(object, NULL);
+}
+
+/* Dump a couple key/object from the cache */
+void stsDumpCache(nmbrString *key, vstring object) {

```

Comment:
suspicious cast to eqFunc in line 235: The signature of this function does not match that of eqFunc: int (eqFunc)(void *, void *)